api/v1alpha4/virtualmachine_affinity_types.go

```diff
@@ -0,0 +1,232 @@
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to zones
+	// that satisfy the anti-affinity expressions specified by this field, but it may choose a zone that
+	// violates one or more of the expressions. The zone that is most preferred is the one with the
+	// greatest sum of weights.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	PreferredDuringSchedulingIgnoredDuringExecution []WeightedZoneSelectorTerm `json:"preferredDuringSchedulingIgnoredDuringExecution,omitempty"`
+}
+
+// VirtualMachineAntiAffinityZoneAffinitySpec defines the anti-affinity scheduling rules
+// related to zones.
+type VirtualMachineAntiAffinityZoneAffinitySpec struct {
+	// +optional
+	// +listType=atomic
+
+	// RequiredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that must be met or the VM will not be scheduled.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	RequiredDuringSchedulingIgnoredDuringExecution []ZoneSelectorTerm `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+
+	// +optional
+	// +listType=atomic
+
+	// PreferredDuringSchedulingIgnoredDuringExecution describes affinity
+	// requirements that should be met, but the VM can still be scheduled if
+	// the requirement cannot be satisfied. The scheduler will prefer to schedule VMs to zones
+	// that satisfy the anti-affinity expressions specified by this field, but it may choose a zone that
+	// violates one or more of the expressions. The node that is most preferred is the one with the
+	// greatest sum of weights.
+	//
+	// When there are multiple elements, the lists of zones corresponding to
+	// each term are intersected, i.e. all terms must be satisfied.
+	PreferredDuringSchedulingIgnoredDuringExecution []WeightedZoneSelectorTerm `json:"preferredDuringSchedulingIgnoredDuringExecution,omitempty"`
+}
+
+// VMAffinityTerm defines the VM affinity/anti-affinity term.
+type VMAffinityTerm struct {
+	// +optional
+
+	// LabelSelector is a label query over a set of VMs.
+	// When omitted, this term matches with no VMs.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// TopologyKey describes where this VM should be co-located (affinity) or not
+	// co-located (anti-affinity). The value is used to match a label on a node, i.e. when
+	// set to topology.kubernetes.io/zone, it means this term applies to nodes
+	// with that label present.
+	TopologyKey string `json:"topologyKey"`
+}
+
+// WeightedVMAffinityTerm defines the weighted VM affinity/anti-affinity term.
```

Comment:
Please drop this and anything else to do with weighting. We agreed with @pranshujain28 that we would punt weighting to some future release.